lib/atomic_tenant/tenantable.rb

```
@@ -0,0 +1,90 @@
+module AtomicTenant::Tenantable
+  extend ActiveSupport::Concern
+
+  @public_tenanted_models = Set.new
+  @private_tenanted_models = Set.new
+
+  class << self
+    attr_reader :public_tenanted_models, :private_tenanted_models
+
+    def register_public_tenanted_model(model)
+      @public_tenanted_models.add(model)
+      @private_tenanted_models.delete(model)
+    end
+
+    def register_private_tenanted_model(model)
+      @private_tenanted_models.add(model)
+      @public_tenanted_models.delete(model)
+    end
+
+    def verify_tenanted(model)
+      query = <<~SQL
+        SELECT relrowsecurity
+        FROM pg_class
+        WHERE relname = $1;
+      SQL
+
+      result = ActiveRecord::Base.connection.exec_query(
+        query,
+        "SQL",
+        [
+          ActiveRecord::Relation::QueryAttribute.new(
+            "relname",
+            model.table_name,
+            ActiveRecord::Type::String.new
+          )
+        ],
+      )
+
+      if !result.first['relrowsecurity']
+        raise "Model #{model.name} is not public but does not have row level security. Did you forget to add row level security in your migration?"
+      end
+    end
+  end
+
+  included do
+    class_attribute :is_tenanted, instance_writer: false, default: true
+
+    before_create :set_tenant_id
+    before_validation :set_tenant_id, on: :create
+    validate :in_current_tenant
```

What happens if you have a record in tenant A, then switch to B, change the tenant_id and try to save the record?  I assume RLS blocks the update?

---

Yeah, RLS will block the update, but I think having a custom validation to provide a nicer error is worth it. That's all this is doing, the query will fail and raise if RLS handles it.

---

Oh, I actually only just understood what you meant. The update will actually silently fail, as postgres essentially adds a `where tenant_id = current_setting('rls.tenant_id', TRUE)` to the query, resulting in it updating 0 rows.